# **The Rise of Face Mask Detection with Convolutional Neural Networks (CNNs)**

## **Background**
The global COVID-19 pandemic has catalyzed notable advancements in computer vision, particularly within the domain of face mask detection. Convolutional Neural Networks (CNNs), a sophisticated form of deep learning architecture, have emerged as an efficacious tool for accurately identifying individuals wearing masks in both images and videos.
### **Prioritizing Public Health**
The capacity to automatically detect face masks serves a pivotal public health function. By deploying CNN-based detection systems in public areas, authorities can:
- Monitor adherence to mask-wearing mandates
- Identify potential health risks
- Enhance contact tracing efforts

### **The Power of CNNs**
CNNs are remarkably proficient at extracting salient features from visual data, such as images. Their architecture, comprising convolutional layers, pooling layers, and fully connected layers, enables them to discern intricate patterns and relationships within the data. In the context of face mask detection, CNNs are trained on extensive datasets of images annotated as "masked" or "unmasked." This training empowers them to effectively distinguish between faces with and without masks in novel images.
### **Key Advantages of CNNs**
CNNs offer several compelling advantages for face mask detection:
- **High Accuracy**: CNNs can achieve impressive accuracy levels, often exceeding 95%, in identifying faces and their mask status.
- **Scalability**: They can be readily scaled to manage large volumes of images and video streams.
- **Real-Time Processing**: Modern CNN implementations facilitate real-time detection, rendering them suitable for time-sensitive applications.

### **Beyond the Pandemic**
The applications of face mask detection using CNNs extend well beyond the immediate public health crisis. This technology holds promise for:
- Access control systems
- Workplace safety monitoring
- Surveillance and security applications

### **The Future of Face Mask Detection**
As CNNs continue to evolve and datasets grow richer, face mask detection is poised to become even more accurate, robust, and versatile. This technology has the potential to play a significant role in safeguarding public health and promoting safety measures across various domains.

In conclusion, the integration of CNN-based face mask detection systems represents a significant stride towards leveraging artificial intelligence for public health and safety. As we look to the future, the continued refinement and application of these technologies promise to enhance our capacity to respond to health crises and ensure secure environments.

## **Problem Statement: Accurate and Efficient Face Mask Detection using Convolutional Neural Networks (CNNs)**
The COVID-19 pandemic has necessitated the widespread adoption of face masks as a critical measure to curb viral transmission. Ensuring consistent and proper mask usage, however, presents a formidable challenge. Manual monitoring is labor-intensive and susceptible to human error. To address this challenge, we propose an automated face mask detection system leveraging Convolutional Neural Networks (CNNs).
### **Challenges:**
- **Accuracy**: The system must reliably distinguish between masked and unmasked individuals, even under varying lighting conditions, poses, and mask types.
- **Real-Time Performance**: For practical deployments in public spaces and surveillance contexts, the system must process video streams efficiently with minimal latency.
- **Scalability**: The system should scale to handle large volumes of images and video data without compromising performance.
- **Generalizability**: The model must generalize across diverse populations and mask designs to ensure broad applicability.

### **Our Approach:**
We propose a CNN-based face mask detection system harnessing the capabilities of deep learning for precise and efficient mask identification. The system will be trained on a meticulously curated dataset, encompassing a wide array of lighting conditions, poses, ethnicities, and mask types. We will employ advanced techniques to optimize the CNN architecture, ensuring real-time performance and scalability while maintaining high accuracy.
### **Significance:**
An accurate and efficient face mask detection system using CNNs can substantially bolster public health efforts by:
- Enhancing adherence to mask-wearing mandates
- Enabling real-time monitoring in public spaces
- Facilitating contact tracing
- Alleviating the burden on manual monitoring personnel

### **Expected Outcomes:**
We aim to develop a CNN-based face mask detection system that achieves:
- **High Accuracy**: Attaining accuracy levels exceeding 95% in identifying masked and unmasked individuals.
- **Real-Time Processing**: Ensuring real-time processing capabilities for video streams.
- **Scalability**: Scaling to support large-scale deployments.
- **Generalizability**: Generalizing effectively to diverse scenarios and mask designs.

# **Importing Essential Libraries**
The following Python libraries are indispensable for this face mask detection project utilizing Convolutional Neural Networks (CNNs):
* **numpy (np):** This library offers robust numerical computing capabilities, which are essential for manipulating and processing image data used in CNN training.
* **pandas (pd):** Pandas provides comprehensive data analysis and manipulation tools, facilitating the handling of annotations or labels associated with the image data.
* **matplotlib (mpl) and matplotlib.pyplot (plt):** These libraries enable sophisticated data visualization, which is crucial for exploring the image dataset, visualizing training progress, and analyzing results.
* **os:** The os library affords interaction with the operating system, which is vital for loading image data from specific directories or file paths.

These libraries constitute the foundational framework for our face mask detection project leveraging CNNs, ensuring efficient data manipulation, analysis, and visualization throughout the development process.

In [1]:
import numpy as np
import pandas as pd
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
import os

# **Leveraging TensorFlow for Deep Learning**
This section introduces the TensorFlow library, an exceptionally powerful framework for building and training deep learning models, including Convolutional Neural Networks (CNNs). In the context of face mask detection:

- **import tensorflow as tf**: This command imports the TensorFlow library and assigns it the alias `tf` for convenience. This alias allows us to access TensorFlow's extensive functionalities seamlessly throughout the code.
- **from tensorflow import keras**: This command imports the Keras API, a high-level interface built atop TensorFlow, designed explicitly for constructing and training neural networks. Keras simplifies the processes of defining, compiling, and training CNN models for face mask detection.

TensorFlow and Keras will serve as the primary deep learning tools for developing and training our CNN-based face mask detection system, ensuring an efficient and streamlined approach to leveraging advanced neural network capabilities.

In [2]:
import tensorflow as tf
from tensorflow import keras

# **Defining Data Directories**
The following code block delineates the directories housing our face mask dataset, segmented for training, validation, and testing purposes:

In [3]:
train_dir = r'D:\sucofindo\pelatihan\CNN\Final_Assignment_Face Mask Dataset\Train'
validation_dir = r'D:\sucofindo\pelatihan\CNN\Final_Assignment_Face Mask Dataset\Validation'
test_dir =r'D:\sucofindo\pelatihan\CNN\Final_Assignment_Face Mask Dataset\Test'

# **Augmenting Data with ImageDataGenerator**
This line of code imports the `ImageDataGenerator` class from the `tensorflow.keras.preprocessing.image` module. In the context of face mask detection using Convolutional Neural Networks (CNNs):
- **ImageDataGenerator**: This class serves as a sophisticated tool for augmenting image datasets. Data augmentation entails the artificial creation of variations in existing images, thereby enriching the training data and enhancing the model's generalization capabilities.

By leveraging various transformations through `ImageDataGenerator`, we can achieve the following:
- **Increase Dataset Size and Diversity**: Augment the training dataset's size and diversity without necessitating additional data collection.
- **Enhance Model Robustness**: Improve the model's robustness by mitigating the risk of overfitting to specific image characteristics.

For instance, `ImageDataGenerator` can apply the following transformations:
- **Random Flips**: Horizontal and vertical flips.
- **Random Rotations**: Rotations by random angles.
- **Random Zooms**: Random zooming in and out.
- **Color Jittering**: Adjustments in brightness, contrast, and saturation.

These transformations enable the model to learn to recognize faces and masks under varied conditions, ultimately leading to improved performance in real-world applications.

In [4]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# **Preparing Training and Validation Data**
This code block demonstrates the process of preparing our training and validation data using ImageDataGenerator for our face mask detection Convolutional Neural Network (CNN) model.

## **Data Generators:**
Two `ImageDataGenerator` objects are instantiated:
- **train_datagen**: Utilized for preprocessing training images.
- **test_datagen**: Utilized for preprocessing validation images, treated similarly to test data in this context.

## **Rescaling:**
Both generators are configured with `rescale=1./255`, normalizing pixel values in images from the 0-255 range to the 0-1 range. This normalization is a common preprocessing step in deep learning models to ensure consistent input values.
## **Flow from Directory:**
- **train_generator**: Created using `train_datagen.flow_from_directory` to read images from the `train_dir` directory.
    - **target_size=(128, 128)**: Resizes images to a uniform size of 128x128 pixels.
    - **batch_size=20**: Groups images into batches of 20 for training the CNN model.
    - **class_mode='binary'**: As we are performing binary classification (masked vs. unmasked), `class_mode` is set to 'binary'.
- **validation_generator**: Created similarly to `train_generator` to read images from the `validation_dir` directory with identical parameters.

In [5]:
train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
        train_dir,
        target_size=(128, 128),
        batch_size=20,
        class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
        validation_dir,
        target_size=(128, 128),
        batch_size=20,
        class_mode='binary')

Found 10000 images belonging to 2 classes.
Found 800 images belonging to 2 classes.


This output confirms that the `flow_from_directory` function successfully identified:

- 10,000 images in the `train_dir` directory, categorized into two classes (likely masked and unmasked).
- 800 images in the `validation_dir` directory, categorized into two classes.

# **Importing Essential Building Blocks: Layers and Models**
This section introduces the fundamental components required for constructing our CNN model for face mask detection:

In [6]:
from tensorflow.keras import layers
from tensorflow.keras import models

- **from tensorflow.keras import layers**: This command imports the `layers` module from TensorFlow's Keras API. This module encompasses a diverse array of pre-defined layers such as convolutional layers, pooling layers, and activation functions, which constitute the essential elements of Convolutional Neural Networks (CNNs).
- **from tensorflow.keras import models**: This command imports the `models` module from TensorFlow's Keras API. This module enables us to define the overall architecture of our CNN model by sequentially stacking the layers imported from the `layers` module.

## **Application in Face Mask Detection:**
In the context of face mask detection, we will utilize various layers from the layers module to perform critical operations, including:
- **Feature Extraction**: Convolutional layers will be employed to extract features from images, identifying patterns such as edges and textures.
- **Dimensionality Reduction**: Pooling layers will be used to reduce the dimensionality of the data, thereby decreasing computational complexity while retaining essential information.
- **Non-Linearity Introduction**: Activation functions will introduce non-linearities, allowing the network to model complex relationships within the data.
- **Classification**: Fully connected layers will classify images as "masked" or "unmasked."

By meticulously combining these layers in a specific architecture using the `models` module, we can construct a robust CNN model capable of accurate face mask detection. This strategic assembly of layers ensures that the model efficiently learns and generalizes from the training data, ultimately enhancing its performance in real-world scenarios.

# **Constructing the CNN Model Architecture**
The following code block outlines the architecture of our Convolutional Neural Network (CNN) model designed for face mask detection. Utilizing the Sequential model API from Keras, we will define the sequence of layers comprising the CNN.

In [7]:
model = models.Sequential()

# First convolutional block
model.add(layers.Conv2D(32, (3, 3), activation='relu',
                        input_shape=(128, 128, 3)))
model.add(layers.MaxPooling2D((2, 2)))

# Second convolutional block
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

# Third convolutional block
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

# Fourth convolutional block
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

# Flatten layer
model.add(layers.Flatten())

# Dense layer for classification
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

c:\Users\hp\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


## **Breakdown of the Architecture:**
- **Sequential Model**: We define a sequential model using `models.Sequential()`. This API allows for the sequential addition of layers to construct the CNN architecture.

- **Convolutional Blocks**: The model incorporates several convolutional blocks, each consisting of:

    - **Convolutional Layer (Conv2D)**: Utilizes filters of size (3, 3) to extract features from the input image. The number of filters (32, 64, 128) determines the number of feature maps learned at each layer.
    - **Activation Layer (ReLU)**: Introduces non-linearity into the network, enabling it to learn more complex patterns.
    - **Pooling Layer (MaxPooling2D)**: Reduces the dimensionality of the data by downsampling the feature maps, thereby enhancing performance and reducing computational cost.
- **Flatten Layer**: This layer transforms the multi-dimensional feature maps into a one-dimensional vector, making the data suitable for the final dense layer.

- **Dense Layer**: This fully connected layer, with 512 neurons and a ReLU activation function, performs the final classification task.

- **Output Layer**: The concluding dense layer contains one neuron with a sigmoid activation function. Given that we are performing binary classification (masked vs. unmasked), the sigmoid function outputs a probability between 0 and 1, where a value closer to 1 indicates "masked" and a value closer to 0 indicates "unmasked."

By meticulously combining these layers in the specified architecture, we construct a robust CNN model capable of accurate face mask detection. This architectural design ensures the efficient learning and generalization from the training data, ultimately enhancing the model's performance in real-world applications.

# **Compiling the CNN Model**
This section elucidates the compilation step for our CNN model, designed for face mask detection. Compilation configures the model's learning process by specifying the loss function, optimizer, and performance metrics to track during training.

In [8]:
from tensorflow.keras import optimizers

model.compile(loss='binary_crossentropy',
              optimizer=optimizers.RMSprop(learning_rate=1e-4),
              metrics=['acc'])

- **Importing the Optimizer**: We import the `optimizers` module from TensorFlow's Keras API to access various optimization algorithms.
- **Compilation Process**: The `model.compile` method configures the training process:
    - **Loss Function (binary_crossentropy)**: Given that we are performing binary classification (masked vs. unmasked), this function measures the discrepancy between the predicted probabilities and the true labels (0 for unmasked, 1 for masked). The model strives to minimize this loss during training.
    - **Optimizer (RMSprop)**: This optimization algorithm adjusts the model's weights based on the calculated loss. In this instance, we employ the RMSprop optimizer with a learning rate of 1×10^−4, a hyperparameter that controls the step size during weight updates.
    - **Metrics (acc)**: We specify 'acc' (accuracy) as the metric to monitor the model's performance during training and validation. Accuracy denotes the proportion of correctly classified images (masked vs. unmasked).

By configuring these parameters, we establish the foundation for the model's learning process, ensuring it is well-equipped to optimize performance and achieve high accuracy in detecting face masks. This meticulous setup is crucial for guiding the model toward effective training and validation, ultimately enhancing its real-world applicability.

# **Training the CNN Model**
This section delineates the training process for our CNN model dedicated to face mask detection. The code trains the model on the prepared training data (`train_generator`) and evaluates its performance on the validation data (`validation_generator`).

In [9]:
history = model.fit(
      train_generator,
      steps_per_epoch=500,
      epochs=20,
      validation_data=validation_generator,
      validation_steps=2)

Epoch 1/20


c:\Users\hp\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


500/500 ━━━━━━━━━━━━━━━━━━━━ 473s 919ms/step - acc: 0.8654 - loss: 0.3129 - val_acc: 1.0000 - val_loss: 0.0261
Epoch 2/20


c:\Users\hp\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:155: UserWarning: Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches. You may need to use the `.repeat()` function when building your dataset.
  self.gen.throw(typ, value, traceback)


500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - acc: 0.0000e+00 - loss: 0.0000e+00 - val_acc: 1.0000 - val_loss: 0.0425
Epoch 3/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 141s 281ms/step - acc: 0.9758 - loss: 0.0738 - val_acc: 1.0000 - val_loss: 0.0256
Epoch 4/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - acc: 0.0000e+00 - loss: 0.0000e+00 - val_acc: 1.0000 - val_loss: 0.0233
Epoch 5/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 126s 251ms/step - acc: 0.9856 - loss: 0.0416 - val_acc: 1.0000 - val_loss: 0.0141
Epoch 6/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 0s 329us/step - acc: 0.0000e+00 - loss: 0.0000e+00 - val_acc: 1.0000 - val_loss: 0.0024
Epoch 7/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 128s 255ms/step - acc: 0.9882 - loss: 0.0318 - val_acc: 1.0000 - val_loss: 0.0065
Epoch 8/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 0s 343us/step - acc: 0.0000e+00 - loss: 0.0000e+00 - val_acc: 1.0000 - val_loss: 0.0019
Epoch 9/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 127s 254ms/step - acc: 0.9909 - loss: 0.0274 - val_acc: 1.0000 - val_loss: 0.0146
Epoch 10/20
500

## **Code Explanation:**
- **model.fit**: This method initiates the training process.
- **train_generator**: Specifies the training data generator that provides batches of images and labels during training.
- **steps_per_epoch=500**: Defines the number of batches to iterate through per epoch, where an epoch represents one complete pass through the training data.
- **epochs=20**: Sets the total number of training epochs.
- **validation_data=validation_generator**: Specifies the validation data generator used to evaluate the model's performance after each epoch.
- **validation_steps=2**: Defines the number of validation batches to use for evaluation during each epoch.

## **Output Interpretation:**
The output showcases the training progress over 20 epochs. Key elements include:
- **Epoch (e.g., Epoch 1/20)**: Indicates the current epoch number out of the total.
- **Progress Within an Epoch (e.g., 500/500)**: Shows progress within an epoch, indicating the current batch being processed (e.g., 500th out of 500 batches).
- **Time per Step (e.g., 919ms/step)**: The average time taken to process each batch.
- **Training Accuracy (acc: 0.8654)**: Represents the percentage of correctly classified images within the current batch.
- **Training Loss (loss: 0.3129)**: Measures how well the model's predictions align with the true labels for the current batch.
- **Validation Accuracy (val_acc: 1.0000)**: Indicates the model's performance on the validation data after each epoch.
- **Validation Loss (val_loss: 0.0261)**: Measures the model's loss on the validation data after each epoch.

By meticulously monitoring these metrics, we can assess the model's learning progress and its ability to generalize to unseen data. The detailed tracking of accuracy and loss, both during training and validation, ensures that the model is effectively learning to distinguish between masked and unmasked faces, thereby enhancing its real-world applicability.

# **Preserving the Trained Model**
The following line of code employs the `model.save` method to preserve the trained CNN model, ensuring its availability for future use.

In [10]:
model.save("model_cnn_project_P1.keras")

- **model.save**: This function, part of the Keras API, facilitates the serialization of the model's architecture, weights, and optimizer configuration into a file.
- **"model_cnn_project_P1.keras"**: This parameter specifies the filename and path for saving the model. Here, the model is saved as "model_cnn_project_P1.keras", with the `.keras` extension, commonly used for Keras models.

## **Benefits of Saving the Model:**
- **Reusability**: Saving the trained model allows for future predictions on new data without the need to retrain, conserving both time and computational resources.
- **Deployment**: For real-world application deployment, the saved model can be loaded into the deployment environment, ready to make predictions.
## **Future Considerations**:
- **Version Control**: Integrating the saved model into a version control system (e.g., Git) enables tracking changes and reverting to previous versions if necessary.
- **File Format**: Depending on deployment requirements and compatibility with other tools, alternative model saving formats such as HDF5 (.h5) or TensorFlow SavedModel (.pb) can be explored.

By preserving the trained model, we ensure its longevity and readiness for various future applications, enhancing its practical utility and deployment efficiency.

# **Cleaning Up Resources After Training**
This section outlines the essential steps to clean up resources following the training of your CNN model for face mask detection.

In [11]:
from tensorflow.keras import backend as K 

K.clear_session()
del model

- **Importing Backend**: We import the `backend` module as `K` from TensorFlow's Keras API. This module provides low-level functionality to interact with the computational backend, such as the CPU or GPU.
- **Clearing the Session (K.clear_session())**: This step is crucial for proper resource management. TensorFlow employs sessions to manage computations and allocate memory for models and data. After training, invoking `K.clear_session()` explicitly terminates the current session, releasing any resources held by the model and backend. This prevents memory leaks and ensures efficient resource utilization, especially when conducting multiple training sessions or working with large models.
- **Deleting the Model (del model)**: This command explicitly deletes the model object from memory. Following the session clearance, this step ensures that all model-related resources are fully released.

## **Importance of Cleanup**:
- **Memory Management**: Clearing the session and deleting the model object aids in preventing memory leaks and ensures efficient memory usage. This is particularly important when dealing with large models or sequentially training multiple models.
- **Resource Management**: TensorFlow sessions can retain computational resources, such as GPUs. Clearing the session allows these resources to be freed and made available for other processes as needed.

By following these cleanup steps, we ensure that our computational resources are managed effectively, promoting optimal performance and preventing unnecessary resource contention in subsequent operations. This meticulous approach is vital for maintaining system stability and efficiency in deep learning workflows.

# **Data Augmentation for Training Robust CNNs**

In [12]:
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,)

test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
        train_dir,
        target_size=(128, 128),
        batch_size=32,
        class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
        validation_dir,
        target_size=(128, 128),
        batch_size=32,
        class_mode='binary')

Found 10000 images belonging to 2 classes.
Found 800 images belonging to 2 classes.


## **Code Explanation:**
### **Augmentation Techniques:**
- **rotation_range=40**: Randomly rotates images by up to 40 degrees.
- **width_shift_range=0.2, height_shift_range=0.2**: Randomly shifts images horizontally or vertically by up to 20% of their width or height, respectively.
- **shear_range=0.2**: Applies a shearing transformation to slightly distort the images.
- **zoom_range=0.2**: Randomly zooms in or out of images by up to 20%.
- **horizontal_flip=True**: Randomly flips images horizontally with a 50% probability.

## **Output Interpretation**
This confirms that the `flow_from_directory` function correctly identified the specified number of images in the training and validation directories (10,000 and 800 images, respectively), distributed across two classes (presumably masked and unmasked).

By employing these data augmentation techniques, we enhance the diversity of our training dataset, thereby improving the robustness and performance of our CNN model in real-world face mask detection scenarios.

## **Model Architecture (using the Sequential API)**
### **Detailed Breakdown:**

**Dropout Layer**: A `Dropout(0.5)` layer is included to randomly drop 50% of the neurons during training, preventing overfitting by reducing dependency on specific neurons.

This comprehensive construction and compilation of the CNN model ensures it is well-prepared for the task of face mask detection, with robust layers for feature extraction, effective regularization techniques, and an efficient optimization strategy.

In [13]:
model = models.Sequential()

# Convolutional Blocks with Max Pooling
model.add(layers.Conv2D(32, (3, 3), activation='relu',
                        input_shape=(128, 128, 3)))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))

# Flatten Layer
model.add(layers.Flatten())

# Dropout for Regularization
model.add(layers.Dropout(0.5))

# Dense Layers for Classification
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

# Model Compilation
model.compile(loss='binary_crossentropy',
              optimizer=optimizers.RMSprop(learning_rate=1e-4),
              metrics=['acc'])

# **Training the CNN Model**

In [14]:
history = model.fit(
      train_generator,
      steps_per_epoch=300,
      epochs=10,
      validation_data=validation_generator,
      validation_steps=20)

Epoch 1/10


c:\Users\hp\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


300/300 ━━━━━━━━━━━━━━━━━━━━ 145s 472ms/step - acc: 0.8166 - loss: 0.4089 - val_acc: 0.9422 - val_loss: 0.1694
Epoch 2/10
 13/300 ━━━━━━━━━━━━━━━━━━━━ 1:49 383ms/step - acc: 0.9137 - loss: 0.2677

c:\Users\hp\AppData\Local\Programs\Python\Python311\Lib\contextlib.py:155: UserWarning: Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches. You may need to use the `.repeat()` function when building your dataset.
  self.gen.throw(typ, value, traceback)


300/300 ━━━━━━━━━━━━━━━━━━━━ 6s 18ms/step - acc: 0.9135 - loss: 0.2546 - val_acc: 0.9438 - val_loss: 0.1572
Epoch 3/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 139s 458ms/step - acc: 0.9105 - loss: 0.2363 - val_acc: 0.9469 - val_loss: 0.1390
Epoch 4/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 6s 17ms/step - acc: 0.9041 - loss: 0.2480 - val_acc: 0.9688 - val_loss: 0.0977
Epoch 5/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 139s 458ms/step - acc: 0.9236 - loss: 0.2035 - val_acc: 0.9703 - val_loss: 0.1053
Epoch 6/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 6s 18ms/step - acc: 0.9499 - loss: 0.1478 - val_acc: 0.9875 - val_loss: 0.0742
Epoch 7/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 141s 466ms/step - acc: 0.9320 - loss: 0.1854 - val_acc: 0.9719 - val_loss: 0.0919
Epoch 8/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 6s 17ms/step - acc: 0.9283 - loss: 0.1741 - val_acc: 0.9625 - val_loss: 0.1271
Epoch 9/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 137s 452ms/step - acc: 0.9368 - loss: 0.1715 - val_acc: 0.9672 - val_loss: 0.1070
Epoch 10/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 6s 17m

# **Leveraging a Pre-trained Model for Feature Extraction**
This section demonstrates the incorporation of a pre-trained convolutional neural network (CNN), specifically VGG19, to enhance our face mask detection model by utilizing its feature extraction capabilities.

In [15]:
# Importing the VGG19 Model Architecture:
from tensorflow.keras.applications import VGG19

#Transfer Learning with VGG19:
conv_base = VGG19(weights='imagenet',
                  include_top=False,
                  input_shape=(128, 128, 3))

## **Code Explanation**
- `VGG19`: This import statement brings the VGG19 model architecture from the Keras applications module into our codebase.
- `VGG19(weights='imagenet', ...)`: This line of code initializes the VGG19 model.
- `weights='imagenet'`: Specifies that the model should load pre-trained weights derived from the ImageNet dataset. These weights encapsulate valuable features learned from a large and diverse set of images.
- `include_top=False`: By setting this parameter to `False`, we exclude the fully connected layers at the top of the VGG19 model, which are typically used for classifying images into 1000 categories. For our face mask detection task, we are interested only in the convolutional layers that perform feature extraction.
- `input_shape=(128, 128, 3)`: Defines the input shape of the images expected by the model, which in this case are 128x128 pixels with three color channels (RGB).

## **Benefits of Transfer Learning**
**Reduced Training Time**: Utilizing the pre-trained weights of VGG19 substantially reduces the training time compared to training a model from scratch. The convolutional layers of VGG19 have already been trained to recognize intricate patterns and features from a vast amount of data, thus providing a solid foundation for our model.

**Improved Performance**: Transfer learning often enhances the model's performance, especially when the source task (ImageNet classification) is closely related to the target task (face mask detection). The pre-trained weights serve as an excellent starting point, allowing the model to focus on learning task-specific features effectively.

By leveraging the pre-trained VGG19 model, we can harness the power of advanced image features learned from the extensive ImageNet dataset, thereby optimizing our face mask detection model for better accuracy and efficiency.

# **Building the Final Classifier on Top of Pre-trained Features**
In this section, we demonstrate the process of constructing a robust classifier by integrating pre-trained features from the VGG19 model for the task of face mask detection.

In [16]:
from tensorflow.keras import models
from tensorflow.keras import layers

model = models.Sequential()
model.add(conv_base) # Pre-trained VGG19 model for feature extraction
model.add(layers.Flatten()) # Flatten the extracted features
model.add(layers.Dense(256, activation='relu')) # Dense layer with ReLU activation
model.add(layers.Dense(1, activation='sigmoid')) # Output layer for binary classification

## **Advantages of this Approach**
**Leveraging Pre-trained Features**: By building upon the VGG19 model, which has been pre-trained on the extensive ImageNet dataset, we harness powerful feature representations that significantly enhance our model's ability to generalize well to new data.

**Efficient Training**: Utilizing a pre-trained model reduces the time and computational resources required for training. The pre-trained convolutional layers already contain valuable information, allowing the model to converge faster and achieve better performance with fewer training epochs.

**Enhanced Performance**: Transfer learning often leads to superior performance in specific tasks, such as face mask detection, as the pre-trained model provides a solid foundation that can be fine-tuned for our target application.

By constructing our classifier on top of the pre-trained VGG19 features, we create a powerful and efficient model capable of accurately distinguishing between masked and unmasked faces.

# **Compiling the Model for Face Mask Detection**
This section details the configuration of the learning process for our Convolutional Neural Network (CNN) model, which has been built upon the pre-trained VGG19 architecture to classify faces as masked or unmasked.

In [17]:
from tensorflow.keras import optimizers

model.compile(loss='binary_crossentropy',
              optimizer=optimizers.RMSprop(learning_rate=2e-5),
              metrics=['acc'])

**Optimizer (RMSprop)**: The `RMSprop` optimizer, imported from Keras optimizers, is selected to fine-tune our model. The optimizer adjusts the model weights during training to minimize the loss function.

## **Advantages of this Compilation Approach**
**Balanced Learning**: The combination of `binary_crossentropy` as the loss function and `RMSprop` with a carefully chosen learning rate ensures a balanced approach to learning. The model can fine-tune its parameters effectively, leveraging the rich feature set provided by the pre-trained VGG19 layers.

**Robust Optimization**: The `RMSprop` optimizer, with its adaptive learning rate, facilitates robust optimization by adjusting the learning rate for each parameter, enhancing the model’s ability to converge to an optimal solution.

**Performance Monitoring**: Utilizing accuracy as the primary metric allows for straightforward and meaningful performance monitoring, ensuring that the model is progressing in the right direction during the training and validation phases.

By meticulously configuring the compilation settings, we set the stage for an effective training process, poised to yield a highly accurate and efficient model for face mask detection.

# **Saving Model Checkpoints During Training**
This section elucidates the process of incorporating a ModelCheckpoint callback to periodically save the model during the training phase of our face mask detection task.

In [18]:
checkpoint_cb = keras.callbacks.ModelCheckpoint("CNN_Final_Project_Model-{epoch:02d}.keras")

**ModelCheckpoint**: This class, part of the Keras callbacks module, facilitates the checkpointing of models during training. It ensures that the model's state is periodically saved to disk.

**Detailed Breakdown**:

- **Callback Creation**:
- `checkpoint_cb = keras.callbacks.ModelCheckpoint("CNN_Final_Project_Model-{epoch:02d}.keras")`: This line instantiates the `ModelCheckpoint` callback. The argument specifies the file path pattern for saving the models.
- **File Path Pattern**:
- **Base Filename**: `CNN_Final_Project_Model-`: This is the base part of the filename for all saved models.
- **Epoch Number Formatting**: `{epoch:02d}`: This placeholder is dynamically replaced with the current epoch number, zero-padded to two digits. This ensures that each saved model file is uniquely named according to its epoch.
- **File Extension**: `.keras`: This denotes the Keras model file format, indicating the saved file contains a serialized Keras model.

## **Benefits of Model Checkpoints**
**Fault Tolerance**:

- **Recovery from Interruptions**: By saving model checkpoints periodically, you can resume training from the latest saved state if the process is interrupted. This is crucial in avoiding the need to restart training from scratch in case of hardware failures or power outages.

**Experiment Comparison**:

- **Performance Evaluation**: Saving models at various points during training (e.g., after each epoch) allows for comparative analysis of their performance on validation data. This aids in selecting the best-performing model by examining different checkpoints.

**Early Stopping Integration**:

- **Optimization**: Model checkpoints can be effectively combined with early stopping techniques. If the validation performance plateaus or deteriorates for a predefined number of epochs, training can be halted, and the best model checkpoint can be loaded for further use. This approach ensures efficient training without overfitting.

# **Implementing in Training Workflow**
Integrating the `ModelCheckpoint` callback into the training process ensures that the model's progress is safeguarded and can be utilized for evaluation and potential continuation. Here's an example of incorporating it into the training loop:

In [19]:
history = model.fit(
      train_generator,
      steps_per_epoch=300,
      epochs=10,
      validation_data=validation_generator,
      validation_steps=20,
      callbacks=[checkpoint_cb])

Epoch 1/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 4063s 14s/step - acc: 0.9239 - loss: 0.1751 - val_acc: 1.0000 - val_loss: 5.6800e-04
Epoch 2/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 177s 544ms/step - acc: 0.9974 - loss: 0.0113 - val_acc: 1.0000 - val_loss: 1.1660e-04
Epoch 3/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 3820s 13s/step - acc: 0.9909 - loss: 0.0304 - val_acc: 0.9984 - val_loss: 0.0038
Epoch 4/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 177s 552ms/step - acc: 0.9953 - loss: 0.0178 - val_acc: 0.9937 - val_loss: 0.0095
Epoch 5/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 3819s 13s/step - acc: 0.9939 - loss: 0.0169 - val_acc: 1.0000 - val_loss: 0.0013
Epoch 6/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 4571s 15s/step - acc: 0.9925 - loss: 0.0312 - val_acc: 1.0000 - val_loss: 0.0039
Epoch 7/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 3872s 13s/step - acc: 0.9953 - loss: 0.0160 - val_acc: 1.0000 - val_loss: 7.7367e-04
Epoch 8/10
300/300 ━━━━━━━━━━━━━━━━━━━━ 180s 559ms/step - acc: 1.0000 - loss: 6.8833e-04 - val_acc: 0.9937 - val_loss: 0.0253
Epoch 9/10
30

In this example, `checkpoint_cb` is passed as an element of the `callbacks` list to the `model.fit` method. This setup ensures that the model's state is saved after each epoch according to the specified file path pattern.

By meticulously integrating model checkpointing, we bolster the training process with robust recovery mechanisms, facilitate comprehensive performance analysis, and optimize training efficiency through early stopping strategies.

# **Evaluating Model Performance on Unseen Data**
This section illustrates the method for assessing the trained Convolutional Neural Network (CNN) model's performance on a separate test dataset, ensuring its efficacy in face mask detection.

In [20]:
test_generator = test_datagen.flow_from_directory(
        test_dir,
        target_size=(128, 128),
        batch_size=32,
        class_mode='binary')

Found 992 images belonging to 2 classes.


This output confirms that the `flow_from_directory` function has successfully located 992 images within the `test_dir` directory, distributed across two classes (presumably 'masked' and 'unmasked'), mirroring the structure of the training and validation datasets.

# **Evaluation Process**
The configured test data generator (`test_generator`) is now poised to evaluate the trained model's performance. This is achieved through the `model.evaluate` function, which computes key metrics such as accuracy and loss on the unseen test data. These metrics provide valuable insights into the model's generalization capabilities, indicating how well it is likely to perform on real-world, unseen data.

In [22]:
model.evaluate(test_generator, steps=31)

31/31 ━━━━━━━━━━━━━━━━━━━━ 99s 3s/step - acc: 0.9988 - loss: 0.0038


[0.0038883087690919638, 0.9989919066429138]

[Test Lost, Test Accuracy]

**Evaluation Method**: The `model.evaluate` function is called with the `test_generator`, which yields the following:

**Test Loss**: A measure of how well the model's predictions align with the true labels, computed on the test dataset.
**Test Accuracy**: The proportion of correctly classified images in the test dataset, reflecting the model's overall performance.

## **Benefits of Evaluation on Unseen Data**
- **Validation of Generalization**: Evaluating on a separate test dataset, which the model has not encountered during training, provides a robust estimate of its generalization ability. This helps in understanding the model's performance in real-world scenarios.
- **Model Improvement**: Insights gained from the evaluation can guide further improvements and refinements to the model, such as adjusting hyperparameters or incorporating additional data augmentation techniques.
- **Confidence in Deployment**: A thorough evaluation ensures that the model is reliable and performant, instilling confidence in its deployment for practical applications.

By rigorously evaluating the CNN model on unseen test data, we ensure its robustness and readiness for real-world face mask detection tasks, ultimately contributing to the development of effective and reliable AI solutions.

# **Conclusion**

This document comprehensively explored the development and evaluation of a Convolutional Neural Network (CNN) model for classifying faces as masked or unmasked. We leveraged techniques like data augmentation and transfer learning to enhance the model's performance.

## **Key Points**:

- The code effectively demonstrates the process of creating data generators for training, validation, and testing data, ensuring proper image preprocessing and batching.
- The model architecture utilizes a pre-trained VGG19 model for feature extraction, followed by a custom head for the binary classification task.
- The model is compiled with an appropriate loss function (binary cross-entropy), optimizer (RMSprop), and metrics (accuracy) for face mask detection.
- Model checkpoints are implemented to save the model state periodically during training, enabling fault tolerance and experiment comparison.
- The test data generator prepares unseen data for evaluating the model's generalizability on a separate dataset.

## **Future Directions**:
- **Data Augmentation and Hyperparameter Tuning**: Experimentation with varied data augmentation techniques and hyperparameter tuning may further enhance the model's performance and resilience.
- **Exploration of Alternative Models**: Investigating other pre-trained models or CNN architectures could potentially yield improved results for this specific dataset, offering a broader spectrum of performance benchmarks.
- **Handling Class Imbalance**: Incorporating techniques to address class imbalance, especially if there is a significant disparity between the number of masked and unmasked images, could lead to more balanced and accurate model predictions.
- **Deployment**: Deploying the trained model to an appropriate platform, such as a mobile application or a web-based interface, will enable real-world face mask detection, extending the model's utility beyond the confines of this study.

By adhering to these strategies, we aim to enhance the effectiveness, reliability, and applicability of the CNN model for face mask detection, ultimately contributing to the development of robust AI solutions for public health and safety.